File: LeetCodePractice/src/math/lexicographical/LexicographicalNumbers.java
package math.lexicographical;
/*Given an integer n, return 1 - n in lexicographical order.
For example, given 13, return: [1,10,11,12,13,2,3,4,5,6,7,8,9].
Please optimize your algorithm to use less time and space. The input size may be as large as 5,000,000.*/
import java.util.*;

public class LexicographicalNumbers {
	public List<Integer> lexicalOrder(int n) {
        List<Integer> res = new ArrayList<>();
        for(int i = 1; i <= 9; i++){
        	dfs(res, i, n);
        }
        return res;
    }
	private void dfs( List<Integer> res, int cur, int n){
		if(cur > n){
			return;
		}
		res.add(cur);
		for(int i = 0; i <= 9; i++){
			dfs(res, cur * 10 + i, n);
		}
	}
}

File: LeetCodePython/LexicographicalNumbers.py
"""
Given an integer n, return all the numbers in the range [1, n] sorted in lexicographical order.

You must write an algorithm that runs in O(n) time and uses O(1) extra space.

Example 1:
Input: n = 13
Output: [1,10,11,12,13,2,3,4,5,6,7,8,9]

Example 2:
Input: n = 2
Output: [1,2]

Constraints:
1 <= n <= 5 * 10^4
"""
from typing import List


class LexicographicalNumbers:
    def lexicalOrder(self, n: int) -> List[int]:
        res = []
        def dfs(cur):
            if cur > n:
                return
            res.append(cur)
            for j in range(10):
                dfs(cur * 10 + j)

        for i in range(1, 10):
            dfs(i)
        return res
